File: align.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <numeric>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using std::make_pair;
using std::make_tuple;

// Not everything needs to be a class
// If you're not using the self variable, that's a hint
std::vector<std::string> readFile(const std::string& fileName, unsigned numLines)
{
    std::vector<std::string> data;
    std::ifstream input(fileName);

    for( unsigned i = 0; i < numLines && input; ++i )
    {
        std::string str;
        std::getline(input, str);
        data.push_back(str);
    }

    return data;
}

void writeData(const std::string& filename, const std::vector<std::string>& lines)
{
    std::ofstream output(filename);
    for( const std::string& l : lines )
    {
        output << l << "\n";
    }
}

typedef std::unordered_set<std::string> Vocabulary;

template<typename Container>
Vocabulary populateVolcabulary(const Container& sentence_list)
{
    Vocabulary result;
    for( const std::string& sentence : sentence_list )
    {
        std::istringstream strm(sentence);
        std::string tmp;
        for( strm >> tmp; strm; strm >> tmp ) // this is roughly equivalent to splitting on spaces
        {
            result.insert(tmp);
        }
    }
    return result;
}

// TODO refactor
// TODO evaluate these hash functions
namespace std {

    template<>
    struct hash<std::tuple<std::string, std::string>>
    {
        std::hash<std::string> str_hash;
        public:
        size_t operator()(const std::tuple<std::string, std::string>& obj) const
        {
            return str_hash(std::get<0>(obj)) ^ str_hash(std::get<1>(obj));
        }
    };

    template<>
    struct hash<std::tuple<unsigned, unsigned>>
    {
        std::hash<unsigned> int_hash;
        public:
        size_t operator()(const std::tuple<unsigned, unsigned>& obj) const
        {
            return int_hash(std::get<0>(obj)) ^ int_hash(std::get<1>(obj));
        }
    };

    template<>
    struct hash<std::tuple<unsigned, unsigned, unsigned, unsigned>>
    {
        std::hash<unsigned> int_hash;
        public:
        size_t operator()(const std::tuple<unsigned, unsigned, unsigned, unsigned>& obj) const
        {
            return int_hash(std::get<0>(obj)) ^ int_hash(std::get<1>(obj))
                ^  int_hash(std::get<2>(obj)) ^ int_hash(std::get<3>(obj));
        }
    };
}



typedef std::unordered_map< std::tuple<std::string, std::string>, double> EmissionParameters;
EmissionParameters initializeEmissionParameters(const Vocabulary& sourceVocab, const Vocabulary& targetVocab)
{
    EmissionParameters params;
    double uniformProbability = 1.0 / targetVocab.size();
    for( const std::string& src : sourceVocab )
    {
        for( const std::string& tgt : targetVocab )
        {
            params.insert(std::make_pair( std::make_tuple(src, tgt), uniformProbability) );
        }
    }
    return params;
}

typedef std::unordered_map<std::tuple<unsigned, unsigned, unsigned, unsigned>, double> AlignmentParameters;
// TODO rename variables
AlignmentParameters initializeAlignmentParameters(
    const std::vector<std::string>& sourceSentenceList,
    const std::vector<std::string>& targetSentenceList
    )
{
    AlignmentParameters alignment_parameters;
    std::unordered_set<std::tuple<unsigned, unsigned>> visited_keys;
    for( unsigned i = 0; i < sourceSentenceList.size() && i < targetSentenceList.size(); ++i )
    {
        const std::string& sourceSentence = sourceSentenceList.at(i);
        const std::string& targetSentence = targetSentenceList.at(i);
        unsigned l = std::count(std::begin(sourceSentence), end(sourceSentence), ' ');
        unsigned m = std::count(std::begin(targetSentence), end(targetSentence), ' ');

        if( visited_keys.find(std::make_tuple(l, m)) != std::end(visited_keys) )
            continue;

        for( unsigned k = 0; k < m; ++k )
        {
            for( unsigned j = 0; j < l; ++j )
            {
                alignment_parameters.insert(make_pair(make_tuple(j, k, l, m), 1.0 / l));
            }
        }

        visited_keys.insert(std::make_tuple(l, m));
    }

    return alignment_parameters;
}

std::vector<std::string> split(const std::string& str, char delim)
{
    std::istringstream strm(str);
    std::vector<std::string> result;
    std::string buff;
    while( std::getline(strm, buff, delim) ) {
        if( buff.size() > 0 )
            result.push_back(buff);
    }
    return result;
}

template<typename T>
static inline T fromString(const std::string& str)
{
    std::istringstream strm(str);
    T val; // can we do this without the default construction?
    strm >> val;
    return val;
}

// TODO this function is really broken
// there are implicit restrictions on type and content of things everywhere
template<typename Container>
double initializeDistanceParameters(const Container& developmentAlignmentsList)
{
    std::unordered_map<std::tuple<std::string, std::string>, unsigned> alignment_count;
    for( const std::string& alignmentLine : developmentAlignmentsList )
    {
        std::istringstream strm(alignmentLine);
        std::string alignment;
        for(strm >> alignment; strm; strm >> alignment) // this is roughly equivalent to splitting on spaces
        {
            std::vector<std::string> alignment_parts;
            if( alignment.size() > 0 && alignment.find('?') != std::string::npos )
                alignment_parts = split(alignment, '?');
            else if( alignment.size() > 0 && alignment.find('-') != std::string::npos )
                alignment_parts = split(alignment, '-');
            else
                throw std::runtime_error("Did not put anything into alignment parts!");
            // Didn't do any bounds checking on alignment parts, so use .at() instead of []
            std::string french_index = alignment_parts.at(0);
            std::string english_index = alignment_parts.at(1);
            // operator[] will default construct a value if it is not already
            // in the set, and the default for unsigned is 0, so this does the right thing(tm)
            alignment_count[make_tuple(french_index, english_index)] += 1;
        }
    }

    // now calculate the distance parameter
    double normalization_constant = std::accumulate(begin(alignment_count), end(alignment_count), 0,
            [](unsigned partial_sum, const std::pair<std::tuple<std::string, std::string>, unsigned> p) {
                return partial_sum + p.second;
            });
    std::vector<int> distance_parameters;
    for( const std::pair<std::tuple<std::string, std::string>, unsigned>& p : alignment_count )
    {
        int distance = abs( fromString<int>(std::get<0>(p.first)) - fromString<int>(std::get<1>(p.first)) );
        if( distance == 0 ) // what's this do?
            distance = 1;
        double probability = static_cast<double>(p.second) / normalization_constant;
        double current_distance_parameter = pow(probability, 1.0 / distance);
        distance_parameters.push_back(current_distance_parameter);
    }

    return std::accumulate(begin(distance_parameters), end(distance_parameters), 0) / distance_parameters.size();
}

struct Parameter
{
    EmissionParameters emit_params;
    AlignmentParameters align_params;
    Parameter(EmissionParameters e, AlignmentParameters a)
        : emit_params(e), align_params(a)
    { }
};

template<typename Container1, typename Container2>
void trainParameters(const Container1& sourceSentenceList, const Container2& targetSentenceList,
        unsigned targetVocabSize,
        EmissionParameters& emission_params, AlignmentParameters& align_params,
        unsigned numIterations, int modelNo
    )
{
    for( unsigned iterations = 0; iterations < numIterations; ++iterations )
    {
        std::unordered_map<int, int> emission_counts;
        std::unordered_map<int, int> alignment_conts;
        std::unordered_set<std::tuple<std::string, std::string>> modified_emission_keys;
        std::unordered_set<std::tuple<unsigned, unsigned, unsigned, unsigned>> modified_alignment_keys;

        for( unsigned current_training_idx = 0; current_training_idx < sourceSentenceList.size(); ++current_training_idx )
        {
            std::vector<std::string> current_source_word_list = split(sourceSentenceList[current_training_idx], ' ');
            std::vector<std::string> current_target_word_list = split(targetSentenceList[current_training_idx], ' ');
            unsigned l = current_source_word_list.size();
            unsigned m = current_target_word_list.size();

            for( unsigned i = 0; i < current_target_word_list.size(); ++i )
            {
                std::string target_word = current_target_word_list[i];
                double denominator = 0;

                if( modelNo == 1 ) // TODO refactor
                {
                    for( const std::string& w : current_source_word_list )
                    {
                        auto tuple_key = make_tuple(w, target_word);
                        EmissionParameters::iterator location = emission_params.find(tuple_key);
                        if( location != end(emission_params) )
                            denominator += location->second;
                        else {
                            denominator += 1.0 / targetVocabSize;
                            emission_params.insert(make_pair(tuple_key, 1.0 / targetVocabSize));
                        }
                    }
                }
                else if( modelNo == 2 )
                {
                    for( unsigned count = 0; count < l; ++count )
                    {
                        auto tuple_key = make_tuple(count, i, l, m);
                        AlignmentParameters::iterator location = align_params.find(tuple_key);
                        double prob = 0;
                        if( location == end(align_params) ) {
                            prob = 1.0 / static_cast<double>(l);
                            align_params.insert(make_pair(tuple_key, prob));
                        }
                        else {
                            prob = location->second;
                        }
                        denominator +=
                               prob * emission_params[make_tuple(current_source_word_list[count], target_word)];
                    }
                }

                for( unsigned j = 0; j < current_source_word_list.size(); ++j )
                {
                    const std::string& source_word = current_source_word_list[j];

                    auto emission_key = make_tuple(source_word, target_word);
                    auto align_key = make_tuple(j, i, l, m);

                    modified_emission_keys.insert(emission_key);
                    modified_alignment_keys.insert(align_key);

                    double delta = emission_params[emission_key];
                    if( modelNo == 2 )
                        delta *= align_params[align_key];
                    delta /= denominator;

                    // The corresponding python (lines 200-227)
                    // is not correct, so I have not transliterated it
                }
            }
        }
    }
}

template<typename Container1, typename Container2>
void makeAndPrintPredictions(const Container1& sourceSentenceList, const Container2& targetSentenceList,
        const EmissionParameters& emissionParams, const AlignmentParameters& alignmentParams,
        double distanceParameter
    )
{
    for( unsigned current_idx = 0; current_idx < 1000; ++current_idx )
    {
        std::vector<std::string> current_source_word_list = split(sourceSentenceList[current_idx], ' ');
        std::vector<std::string> current_target_word_list = split(targetSentenceList[current_idx], ' ');
        unsigned l = current_source_word_list.size();
        unsigned m = current_target_word_list.size();

        std::ostringstream alignmentString;
        for(unsigned i = 0; i < current_target_word_list.size(); ++i )
        {
            const std::string& target_word = current_target_word_list[i];
            alignmentString << i;

            unsigned current_max_index = 0;
            double current_max;

            for( unsigned j = 0; j < current_source_word_list.size(); ++j)
            {
                const std::string& source_word = current_source_word_list[j];
                double current_val =
                      emissionParams.at(make_tuple(source_word, target_word))
                    * alignmentParams.at(make_tuple(j,i,l,m))
                    * pow(distanceParameter, abs(i-j));
                if( current_max < current_val )
                {
                    current_max = current_val;
                    current_max_index = j;
                }
            }
            std::cout << alignmentString.str() << "\n";
        }
    }
}

int main()
{
    std::vector<std::string> englishSentenceList = readFile("data/hansards.e", 9999);
    std::vector<std::string> frenchSentenceList = readFile("data/hansards.f", 9999);
    std::vector<std::string> developmentAlignmentsList = readFile("data/hansards.a", 37);

    // compute the french vocabulary
    std::unordered_set<std::string> frenchVocabulary = populateVolcabulary(frenchSentenceList);

    // get the initial parameters
    double distanceParameter = initializeDistanceParameters(developmentAlignmentsList);

    // Create the parameters for training
    EmissionParameters emission_parameters;
    AlignmentParameters alignment_parameters;

    // train for model 1
    trainParameters(englishSentenceList, frenchSentenceList, frenchVocabulary.size(), emission_parameters, alignment_parameters, 10, 1);

    // train for model 2
    trainParameters(englishSentenceList, frenchSentenceList, frenchVocabulary.size(), emission_parameters, alignment_parameters, 10, 2);

    // Write the predictions out
    makeAndPrintPredictions(englishSentenceList, frenchSentenceList, emission_parameters, alignment_parameters, distanceParameter);

    return 0;
}

File: align.py
#! /usr/bin/env python2


xrange = range


class Sentence:
    def __init__(self, text):
        self.words = text.split()

    def __getitem__(self, index):
        return self.words[index]

    def __iter__(self):
        return self.words.__iter__()

    def __len__(self):
        return len(self.words)


class SentencePair:
    def __init__(self, first, second):
        self.source = first
        self.target = second
        self.alignment = {}
        for k in xrange(0, len(second)):
            for j in xrange(0, len(first)):
                self.alignment[(j, k)] = 1 / float(len(first))


def readFile(fileName, numLines):
    data = []
    i = 0
    with open(fileName) as fileObject:
        for line in fileObject:
            data.append(line)
            i = i + 1
            if i == numLines:
                break

    return data


def readSentenceFile(fileName, numLines):
    return [Sentence(l) for l in readFile(fileName, numLines)]


def writeData(fileName, data):
    with open(fileName) as fileObject:
        fileObject.write(data + '\n')


def populateVocabulary(sentenceList):
    vocabulary = set()
    for sentence in sentenceList:
        for word in sentence:
            vocabulary.add(word)
    return vocabulary


# Tracks the probability that a given source word emits
# a particular target word, regardless of context
class EmissionParameters:
    def __init__(self, sourceVocabulary, targetVocabulary):
        self._changedParameters = {}
        print("initializing " + str(len(sourceVocabulary) * len(targetVocabulary)) + " parameters")
        self._uniformProbability = 1 / float(len(targetVocabulary))

    def setProbability(self, source, target, prob):
        self._changedParameters[(source, target)] = prob

    def getProbability(self, source, target):
        if (source, target) in self._changedParameters:
            return self._changedParameters[(source, target)]
        else:
            return self._uniformProbability


# Takes in pre-aligned data, measures how far apart the corresponding words
# are from each other.
def initializeDistanceParameters(developmentAlignmentsList):
    # Map from aligned (target index, source index) -> ocurrences in the supervised training set
    # alignment count (target index, source index) is the number of times that the word at the target index is
    # aligned to the word at the source index in the supervised set
    alignmentCount = {}

    for alignmentLine in developmentAlignmentsList:
        alignments = alignmentLine.strip().split()
        for alignment in alignments:
            alignmentParts = []
            # '?' means uncertain alignment
            # '-' means certain
            if '?' in alignment:
                alignmentParts = alignment.split('?')
            elif '-' in alignment:
                alignmentParts = alignment.split('-')
            frenchIndex = alignmentParts[0]
            englishIndex = alignmentParts[1]
            if (frenchIndex, englishIndex) not in alignmentCount:
                alignmentCount[(frenchIndex, englishIndex)] = 0
            alignmentCount[(frenchIndex, englishIndex)] += 1

    # now calculate the distance parameter
    # distance parameter is k s.t. k^|i-j| = prob(alignment distance)
    # so we go over the data set and compute k for each distance,
    # then take the weighed average over the alignments
    # This is possibly not the best way to estimate k
    normalizationConstant = sum(alignmentCount.values())
    DistanceParameters = []
    for key in alignmentCount.keys():
        distance = abs(int(key[0]) - int(key[1]))
        if distance == 0:
            distance = 1
        probability = float(alignmentCount[key]) / float(normalizationConstant)
        currentDistanceParameter = pow(probability, float(1) / float(distance))
        DistanceParameters.append(currentDistanceParameter)

    return float(sum(DistanceParameters)) / float(len(DistanceParameters))


def incrementOrInsert(dictionary, key, amount):
    try:
        dictionary[key] += amount
    except KeyError:
        dictionary[key] = amount

# this class trains the alignment parameters using EM.
def trainParameters(sentencePairs, frenchVocabSize, emissionParameters, numIterations, modelNo):
    for iteration in xrange(0, numIterations):
        print('iteration = ' + str(iteration))
        emissionCounts = {}
        alignmentCounts = {}
        modifiedEmissionKeys = []
        modifiedAlignmentKeys = []

        for pair in sentencePairs:
            l = len(pair.source)
            m = len(pair.target)

            for i in xrange(0, len(pair.target)):
                targetWord = pair.target[i]
                denominator = 0

                if modelNo == 1:
                    for w in pair.source:
                        denominator += emissionParameters.getProbability(w, targetWord)

                elif modelNo == 2:
                    for count in xrange(0, len(pair.source)):
                        try:
                            #print('pair.alignment is ' + str(pair.alignment))
                            denominator = denominator + (pair.alignment[(count, i)] *
                                    emissionParameters.getProbability(pair.source[count], targetWord))
                        except KeyError:
                            pair.alignment[(count, i)] = 1 / float(len(pair.source))
                            denominator += (pair.alignment[(count, i)] *
                                    emissionParameters[(pair.source[count], targetWord)])

                for j in xrange(0, len(pair.source)):
                    sourceWord = pair.source[j]
                    modifiedEmissionKeys.append((sourceWord, targetWord))
                    modifiedAlignmentKeys.append((j, i, pair))

                    # calculate the delta parameter
                    delta = emissionParameters.getProbability(sourceWord, targetWord)

                    if modelNo == 2:
                        delta = delta * pair.alignment[(j, i)]

                    delta = float(delta) / float(denominator)

                    # update the counts
                    incrementOrInsert(emissionCounts, (sourceWord, targetWord), delta)
                    incrementOrInsert(emissionCounts, sourceWord, delta)
                    incrementOrInsert(alignmentCounts, (j, i, pair), delta)
                    incrementOrInsert(alignmentCounts, (i, pair), delta)

        # update the parameters
        for key in modifiedEmissionKeys:
            emissionParameters.setProbability(key[0], key[1], float(emissionCounts[key]) / float(emissionCounts[key[0]]))

        if modelNo == 2:
            for key in modifiedAlignmentKeys:
                key[2].alignment[(key[0], key[1])] = float(alignmentCounts[key]) / float(alignmentCounts[(key[1], key[2])])


# this class predicts the alignments and prints them to standard out.
def makeAndPrintPredictions(sentencePairs,
                            emissionParameters,
                            distanceParameter):
    for pair in sentencePairs:
        alignmentString = ''
        for i in xrange(0, len(pair.target)):
            targetWord = pair.target[i]
            alignmentString = alignmentString + str(i)
            currentMaximumIndex = 0
            currentMaximum = 0.0
            for j in xrange(0, len(pair.source)):
                sourceWord = pair.source[j]
                if currentMaximum < emissionParameters.getProbability(sourceWord, targetWord) * pair.alignment[j, i] * pow(distanceParameter, abs(i - j)):
                    currentMaximum = emissionParameters.getProbability(sourceWord, targetWord) * pair.alignment[j, i] * pow(distanceParameter, abs(i - j))
                    currentMaximumIndex = j
            alignmentString = alignmentString + '-' + str(currentMaximumIndex) + ' '

        alignmentString = alignmentString[0:len(alignmentString)-1]
        print(alignmentString)


def main():
    # read the files
    englishSentenceList = readSentenceFile('data/hansards.e', 100)
    frenchSentenceList = readSentenceFile('data/hansards.f', 100)
    developmentAlignmentsList = readFile('data/hansards.a', 37)

    # get the french vocab
    englishVocabulary = populateVocabulary(englishSentenceList)
    frenchVocabulary = populateVocabulary(frenchSentenceList)

    # get the initial parameters
    distanceParameter = initializeDistanceParameters(developmentAlignmentsList)

    # create the parameters for training
    emissionParameters = EmissionParameters(englishVocabulary, frenchVocabulary)
    #alignmentParameters = AlignmentParameters(englishSentenceList, frenchSentenceList)
    sentencePairs = [SentencePair(x, y) for (x, y) in zip(englishSentenceList, frenchSentenceList)]
    frenchVocabSize = len(frenchVocabulary)

    # train the EM parameters for model 1
    trainParameters(sentencePairs, frenchVocabSize,
                    emissionParameters, 10, 1)

    # train the EM parameters for model 2
    trainParameters(sentencePairs, frenchVocabSize,
                    emissionParameters, 10, 2)

    # write the predictions to file
    makeAndPrintPredictions(sentencePairs,
                            emissionParameters,
                            distanceParameter)

main()
